adam/src/main/kotlin/com/malinskiy/adam/request/shell/v2/SyncShellCommandRequest.kt

```diff
@@ -79,4 +72,25 @@ abstract class SyncShellCommandRequest<T : Any?>(val cmd: String, target: Target
 
         return convertResult(shellCommandResult)
     }
+
+    private suspend fun ByteArrayDataOutput.write(socket: Socket): Unit {
+        val length = socket.readIntLittleEndian()
+        if (length > buffer.size) { // Grow buffer as needed.
+            buffer = ByteArray(nextDataSize(length))
```

Comment:
This can potentially allocate 2Gb of heap space per request. I seriously doubt this is what you want.

Reply:
How do you reason it would grow this large?

Reply:
This would happen when the ADB server sends you max int here

Reply:
Oh wow. Is this a bug in ADB or considered normal behavior? Any additional info you have would help.

BTW, I have no problem making the changes, just want to understand the issue better.

Reply:
It's a side effect of using a non-documented and non-standardized protocol. Although there is logic in the C implementation of ADB server for max packet size the protocol here supports sending max int. In any case most of the protocols that do this have to frame the stream hence its much better to just read the available portion of the socket stream and back pressure the rest via not reading.

Reply:
How do I know when I have come to the end if I get max int?

Reply:
Have a counter initialised to max int, decrease it by the amount you read and loop while left to read is > 0
The sample for streaming channel request is already doing that, I've linked it above